packages/flutter_tools/lib/src/compute_dev_dependencies.dart

```diff
@@ -2,126 +2,162 @@
+    if (result.containsKey(current)) {
+      continue;
+    }
+    final _PackageDetails? details = _loadPackageDetails(
+      current,
+      packageConfig,
+      project,
+      fileSystem,
+    );
+    if (details == null) {
+      continue;
+    }
+
+    packageNamesToVisit.addAll(details.dependencies);
+    result[current] = Dependency(
+      details.pubspec,
+      details.dependencies,
+      details.rootUri,
+      isExclusiveDevDependency: true,
     );
   }
 
-  List<T> asListOrFail<T>(Object? value, String name) {
-    // Allow omitting a list as empty to default to an empty list
-    if (value == null) {
-      return <T>[];
+  // Do a second traversal of only the non-dev-dependencies, to patch up the
+  // `isExclusiveDevDependency` property.
+  final Set<String> visitedDependencies = <String>{};
+  packageNamesToVisit.add(project.manifest.appName);
+  while (packageNamesToVisit.isNotEmpty) {
+    final String current = packageNamesToVisit.removeLast();
+    if (!visitedDependencies.add(current)) {
+      continue;
     }
-    if (value is! List<Object?>) {
-      fail('Expected field "$name" to be a list, got "$value"');
+    final Dependency? currentDependency = result[current];
+    if (currentDependency == null) {
+      continue;
     }
-    return <T>[
-      for (final Object? any in value)
-        if (any is T) any else fail('Expected element to be a $T, got "$any"'),
-    ];
+    result[current] = Dependency(
+      currentDependency.pubspec,
+      currentDependency.dependencies,
+      currentDependency.rootUri,
+      isExclusiveDevDependency: false,
+    );
+    packageNamesToVisit.addAll(currentDependency.dependencies);
   }
+  return result;
+}
 
-  // Parse the JSON roughly in the following format:
-  //
-  // ```json
-  // {
-  //   "root": "my_app",
-  //   "packages": [
-  //     {
-  //       "name": "my_app",
-  //       "kind": "root",
-  //       "dependencies": [
-  //         "foo_plugin",
-  //         "bar_plugin"
-  //       ],
-  //       "directDependencies": [
-  //         "foo_plugin"
-  //       ],
-  //       "devDependencies": [
-  //         "bar_plugin"
-  //       ]
-  //     }
-  //   ]
-  // }
-  // ```
-  final List<Map<String, Object?>> packages = asListOrFail(jsonResult['packages'], 'packages');
+typedef _PackageDetails = ({List<String> dependencies, YamlMap pubspec, Uri rootUri});
 
-  Map<String, Object?> packageWhere(
-    bool Function(Map<String, Object?>) test, {
-    required String reason,
-  }) {
-    return packages.firstWhere(test, orElse: () => fail(reason));
+/// Attempts to load the pubspec.yaml for [packageName] and extract the list of
+/// dependencies.
+///
+/// Returns `null` on failure (malformed pubspec etc.).
+_PackageDetails? _loadPackageDetails(
+  String packageName,
+  PackageConfig packageConfig,
+  FlutterProject project,
+  FileSystem fileSystem,
+) {
+  final Package? package = packageConfig[packageName];
+  if (package == null) {
+    return null;
```

Comment:
Is it worth adding verbose logging with the reason we failed to load package details where we return null?

Reply:
Yeah - that is a good idea!

Added `printTrace` statements